### PR TITLE
schedule: introduce the waiting operators

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
Part of #1304, but rewrite the logic.

### What is changed and how it works?
This PR introduces the waiting operators to solve two problem:

1. limit the number of operator created by each scheduler
2. try to be fair to each scheduler

### Check List 

Tests 

 - Unit test

Code changes

 - Has configuration change

Related changes

 - Need to update the documentation
 - Need to be included in the release notes

### CODE REVIEW

File: server/schedule/operator_controller.go [hunk 1]
```diff
@@ -70,31 +80,75 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo) {
...
+	if oc.swo[desc] == nil {
+		oc.swo[desc] = &SchedulerStatus{0, oc.cluster.GetSchedulerMaxWaitingOperator()}
+	}
+	if oc.swo[desc].current >= oc.swo[desc].max {
+		oc.Unlock()
+		return false
+	}
+	oc.wop.PutOperator(op)
+	oc.swo[desc].current++
+	if desc == "merge-region" {
```
Comment (disksing) [inline server/schedule/operator_controller.go]: How about check `if len(ops) > 1` instead?

File: server/schedule/operator_controller.go [hunk 2]
```diff
@@ -70,31 +80,75 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo) {
...
+		if op == nil {
+			return
+		}
+		ops = append(ops, op)
+		if op.Desc() == "merge-region" {
+			ops = append(ops, oc.wop.GetOperator())
+		}
+		for _, op := range ops {
+			if !oc.checkAddOperator(op) {
+				operatorCounter.WithLabelValues(op.Desc(), "canceled").Inc()
```
Comment (disksing) [inline server/schedule/operator_controller.go]: Need to decrease `current` when cancel add operator?

**Comment by nolouch** [comment]
Maybe we can try this http://utopia.duth.gr/~pefraimi/research/data/2007EncOfAlg.pdf ?

File: server/schedule/operator_controller.go [hunk 3]
```diff
@@ -166,21 +178,68 @@ func (oc *OperatorController) PushOperators() {
...
+		oc.Unlock()
 		return false
 	}
-	for _, op := range ops {
-		oc.addOperatorLocked(op)
+
+	op := ops[0]
+	desc := op.Desc()
+	if oc.wopStatus[desc] == nil {
+		oc.wopStatus[desc] = &WaitingOperatorStatus{0, oc.cluster.GetSchedulerMaxWaitingOperator()}
```
Comment (nolouch) [inline server/schedule/operator_controller.go]: Could't it take effect as a dynamic configuration?

File: server/schedule/operator_controller.go [hunk 4]
```diff
@@ -166,21 +178,68 @@ func (oc *OperatorController) PushOperators() {
...
+	}
+	if oc.wopStatus[desc].current >= oc.wopStatus[desc].max {
+		oc.Unlock()
+		return false
```
Comment (nolouch) [inline server/schedule/operator_controller.go]: need metrics?

File: server/schedule/operator_controller.go [hunk 5]
```diff
@@ -166,21 +178,68 @@ func (oc *OperatorController) PushOperators() {
...
+	defer oc.Unlock()
+	var ops []*Operator
+	for {
+		ops = ops[:0]
+		op := oc.wop.GetOperator()
+		if op == nil {
+			return
+		}
+		ops = append(ops, op)
+		if op.Desc() == "merge-region" {
```
Comment (nolouch) [inline server/schedule/operator_controller.go]: the `GetOperator` cannot get the corresponding `merge operator` in second times.

File: server/schedule/waiting_operator.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+// limitations under the License.
+
+package schedule
+
+import (
+	"math/rand"
+	"time"
+)
+
+// weight is used to represent the weight of different priorities of operators.
```
Comment (shafreeck) [inline server/schedule/waiting_operator.go]: weight -> Weight

File: server/schedule/waiting_operator.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	ops    []*Operator
+}
+
+// RandBuckets is an implementation of waiting operators
+type RandBuckets struct {
+	totalWeight float64
+	buckets     []*Bucket
+}
+
+// NewRandBuckets create a random buckets.
```
Comment (shafreeck) [inline server/schedule/waiting_operator.go]: create -> creates

File: server/schedule/waiting_operator.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"math/rand"
+	"time"
+)
+
+// weight is used to represent the weight of different priorities of operators.
+var weight = []float64{1.0, 4.0, 9.0}
+
+// WaitingOperator is an interface of waiting operator.
```
Comment (shafreeck) [inline server/schedule/waiting_operator.go]: waiting operator -> waiting operators

File: server/schedule/waiting_operator.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	bucket.ops = append(bucket.ops, op)
+}
+
+// GetOperator gets an operator from the random buckets.
+func (b *RandBuckets) GetOperator() *Operator {
+	if b.totalWeight == 0 {
+		return nil
+	}
+	rand.Seed(time.Now().UnixNano())
```
Comment (shafreeck) [inline server/schedule/waiting_operator.go]: Call `rand.Seed` once when init is enough. Maybe put it in an `init` function?

File: server/schedule/operator_controller.go [hunk 10]
```diff
@@ -51,6 +51,12 @@ type HeartbeatStreams interface {
 	SendMsg(region *core.RegionInfo, msg *pdpb.RegionHeartbeatResponse)
 }
 
+// WaitingOperatorStatus is used to limit the count of each kind of operators.
```
Comment (shafreeck) [inline server/schedule/operator_controller.go]: Why do not define it in `waiting_operator.go`? I guess it is specific to waiting operators.

File: server/schedule/operator_controller.go [hunk 11]
```diff
@@ -166,21 +178,68 @@ func (oc *OperatorController) PushOperators() {
...
+	desc := op.Desc()
+	if oc.wopStatus[desc] == nil {
+		oc.wopStatus[desc] = &WaitingOperatorStatus{0, oc.cluster.GetSchedulerMaxWaitingOperator()}
+	}
+	if oc.wopStatus[desc].current >= oc.wopStatus[desc].max {
+		oc.Unlock()
+		return false
+	}
+	oc.wop.PutOperator(op)
+	if len(ops) > 1 {
```
Comment (shafreeck) [inline server/schedule/operator_controller.go]: I know this is the merge-region case, however, I think it is better to add some comments here which is friendly to the community developers.

File: server/schedule/operator_controller.go [hunk 12]
```diff
@@ -183,6 +219,34 @@ func (oc *OperatorController) AddOperator(ops ...*Operator) bool {
...
+			}
+			oc.wopStatus.ops[ops[0].Desc()]--
+			continue
+		}
+		oc.wopStatus.ops[ops[0].Desc()]--
+		break
+	}
+
+	for _, op := range ops {
+		oc.addOperatorLocked(op)
```
Comment (nolouch) [inline server/schedule/operator_controller.go]: The operator always nil after breaking from the loop.

File: server/schedule/waiting_operator.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	}
+	r := rand.Float64()
+	var sum float64
+	for _, bucket := range b.buckets {
+		if len(bucket.ops) == 0 {
+			continue
+		}
+		proportion := bucket.weight / b.totalWeight
+		if r >= sum && r < sum+proportion {
+			if len(bucket.ops) == 1 {
```
Comment (nolouch) [inline server/schedule/waiting_operator.go]: If `merge-region` in the queue head, it no enter this logic, and the `weight` does not clear.

**Pull request merged**